lib/puppet/provider/selinux_port/semanage.rb

```diff
@@ -0,0 +1,77 @@
+Puppet::Type.type(:selinux_port).provide(:semanage) do
+  desc 'Support managing SELinux custom port definitions via semanage'
+
+  defaultfor kernel: 'Linux'
+
+  commands semanage: 'semanage'
+
+  mk_resource_methods
+
+  def self.parse_semanage_lines(lines, source)
+    ret = {}
+    lines.each do |line|
+      split = line.split(%r{\s+})
+      context = split.shift
+      protocol = split.shift
+      # The port-range list is comma-separated
+      ports = split.map { |range| range.delete(',') }
+      # interestingly enough, it is completely valid to define a port that overlaps with a port-range,
+      # but if an existing *exact* definition exists, that causes problems
+      ports.each do |port|
+        ret[port] = {
+          ensure: :present,
+          name: "#{protocol}_#{port}",
+          context: context,
+          ports: port,
+          protocol: protocol.to_sym,
+          source: source
+        }
+      end
+    end
+    ret
+  end
+
+  def self.instances
+    # no way to do this with one call as far as I know
+    custom = parse_semanage_lines(semanage('port', '-lC', '-n').split("\n"), 'custom')
```

Comment:
I think the term `custom` should be called `local`. all the config-files written be semanage have the `*.local` suffix.

Caution: `semanage port` operations fail when selinux is disabled. You can add a port with `--noreload` but `--list` fails.